app/earn/[...vault]/portfolio/portfolio-chart.tsx

```diff
@@ -0,0 +1,119 @@
+'use client'
+
+import { useLayoutEffect, useRef, useState } from 'react'
+import { Spacer, Tooltip } from '@nextui-org/react'
+import { useVaultDeposit } from '../hooks/use-vault-deposit-change'
+import { useVaultContext } from '../hooks/use-vault-context'
+import styles from './portfolio-chart.module.scss'
+import { useChart } from './use-chart'
+import { ChangeIndicator } from './components/change-indicator'
+import type { Vault } from '@/api'
+import { toStringNumberFromDecimals, toUSDValue } from '@/shared'
+import CompactNumber from '@/components/compact-number/compact-number'
+import { Row } from '@/components/row/Row'
+
+interface Props {
+  size: number
+  proportion: number
+}
+
+export function PortfolioChart({ size, proportion }: Props) {
+  useChart({
+    size,
+    lineWidth: 10,
+    animationStep: 0.01,
+    value: +proportion.toFixed(2),
+  })
+
+  return (
+    <div className={styles.container}>
+      <AmountChangeInfo size={size} />
+      <canvas id="portfolio-chart" width={size} height={size} />
+    </div>
+  )
+}
+
+function AmountChangeInfo({ size }: Omit<Props, 'proportion'>) {
+  const ref = useRef<HTMLDivElement>(null)
+
+  const [scale, setScale] = useState(1)
+
+  const { vault } = useVaultContext()
+  const [total, change] = useVaultDeposit()
+  const { decimals, price, symbol: assetSymbol } = vault.asset
+  const { value: tokenPrice, decimals: priceDecimals } = price
+  const priceInUSD = toUSDValue(total, decimals, tokenPrice)
+
+  useLayoutEffect(() => {
+    const { current } = ref
+    if (!current)
+      return
+
+    let maxWidth = 0
+    for (const child of current.children) maxWidth = Math.max(maxWidth, (child as HTMLElement).offsetWidth)
+
+    const scale = Math.min((size * 0.6) / maxWidth, 1)
+    setScale(scale)
+  }, [size, total, change])
+
+  return (
+    <Tooltip content={<TooltipContent />}>
+      <div className={styles.info} ref={ref} style={{ transform: `scale(${scale})` }}>
+        <TokenAmount vault={vault} value={total} change={change} decimals={decimals} />
+        <Price value={priceInUSD} decimals={priceDecimals} />
+      </div>
+    </Tooltip>
+  )
+
+  function TooltipContent() {
+    const valueAccurate = toStringNumberFromDecimals(total, decimals)
+    const valueUSDAccurate = toStringNumberFromDecimals(priceInUSD, priceDecimals)
+    return (
+      <>
+        <Row justify="center">
+          {valueAccurate}&nbsp;{assetSymbol}
+        </Row>
+        <Spacer y={0.5} />
+        <Row justify="center">${valueUSDAccurate}</Row>
+      </>
+    )
+  }
```

Comment:
Function inside a function? It can just pass data through props, instead of context